### PR TITLE
Cubeprop improvements

### PR DESCRIPTION
**Pull request by PeterKraus**
## Description
Added a couple of new features to cubeprop. The functionality seems to work, but the PR is rather unpolished. An illustration of me while writing this PR can be found here:

https://ih0.redbubble.net/image.511550362.9563/flat,1000x1000,075,f.u1.jpg

## Todos
Notable points that this PR has either accomplished or will accomplish.
* **Developer Interest**
  - [x] Implements printing of ECP electrons, issue #1136 
* **User-Facing for Release Notes**
  - [x] Added keyword `"FRONTIER_ORBITALS"` to cubeprop tasks. This computes the alpha HOMO and LUMO (if nalpha == nbeta), or the highest-energy doubly occupied (DOMO), the singly occupied (SOMO) and the unoccupied (LUMO) orbitals of both spins.
  - [x] Added keyword `"DUAL_DESCRIPTOR"` to cubeprop tasks. Computes (for closed shell systems) the dual descriptor function (ie. the difference of the Fukui functions) from the HOMO and LUMO. Some references: [10.1021/jp046577a](http://dx.doi.org/10.1021/jp046577a) and [10.1007/s10910-014-0437-7](http://dx.doi.org/10.1007/s10910-014-0437-7).

## Checklist
- [x] Add documentation of the new features to the manual.
- [X] Add tests
- [X] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

File: psi4/src/psi4/libcubeprop/csg.cc [hunk 1]
```diff
@@ -695,4 +722,22 @@ std::pair<double, double> CubicScalarGrid::compute_isocontour_range(double* v2,
...
+    std::stringstream ecp_head;
+    ecp_head.str("");
+    if (primary_->has_ECP()) {
+        ecp_head << " Total core: " << primary_->n_ecp_core() << " [e] from 1-indexed atoms (";
+        std::stringstream ecp_atoms;
+        std::stringstream ecp_ncore;
+        for (int A = 0; A < mol_->natom(); A++) {
+            if (primary_->n_ecp_core(mol_->label(A)) > 0) {
+                ecp_atoms << A+1 << "[" << mol_->symbol(A).c_str() << "], ";
+                ecp_ncore << primary_->n_ecp_core(mol_->label(A)) << ", ";
```
Comment (loriab) [inline psi4/src/psi4/libcubeprop/csg.cc]: Nice & correct use of `label` vs. `symbol` here. c++ folks correct me if I'm wrong, but I don't think the `c_str()` is needed for `mol_->symbol(A).c_str()`.

File: psi4/src/psi4/libcubeprop/csg.cc [hunk 2]
```diff
@@ -649,6 +649,33 @@ void CubicScalarGrid::compute_orbitals(std::shared_ptr<Matrix> C, const std::vec
...
 }
+void CubicScalarGrid::compute_difference(std::shared_ptr<Matrix> C, const std::vector<int>& indices,
+                                         const std::string& label, bool square, const std::string& type) {
+    auto C2 = std::make_shared<Matrix>(primary_->nbf(), indices.size());
+    double** Cp = C->pointer();
+    double** C2p = C2->pointer();
+    for (int k = 0; k < indices.size(); k++) {
+        C_DCOPY(primary_->nbf(), &Cp[0][indices[k]], C->colspi()[0], &C2p[0][k], C2->colspi()[0]);
+    }
+    double** v_t = block_matrix(indices.size(), npoints_);
```
Comment (dgasmith) [inline psi4/src/psi4/libcubeprop/csg.cc]: Can you switch this to a `auto v_t = std::make_shared<Matrix>(indices.size(), npoints_)` and a vector below. Trying to discourage `new/delete/free` in general.

File: psi4/src/psi4/libcubeprop/csg.cc [hunk 3]
```diff
@@ -649,6 +649,33 @@ void CubicScalarGrid::compute_orbitals(std::shared_ptr<Matrix> C, const std::vec
     }
     free_block(v);
 }
+void CubicScalarGrid::compute_difference(std::shared_ptr<Matrix> C, const std::vector<int>& indices,
+                                         const std::string& label, bool square, const std::string& type) {
+    auto C2 = std::make_shared<Matrix>(primary_->nbf(), indices.size());
+    double** Cp = C->pointer();
+    double** C2p = C2->pointer();
```
Comment (dgasmith) [inline psi4/src/psi4/libcubeprop/csg.cc]: Check that `C->nirrep() == 1`, so you do not get odd failures for matrices with symmetry.

File: psi4/src/psi4/libcubeprop/csg.cc (hunk 1 above)
Comment (dgasmith) [reply psi4/src/psi4/libcubeprop/csg.cc]: Correct.

File: psi4/src/psi4/libcubeprop/cubeprop.cc [hunk 4]
```diff
@@ -154,6 +158,52 @@ void CubeProperties::raw_compute_properties() {
...
             if (indsb0.size()) compute_orbitals(Cb_, indsb0, labelsb, "Psi_b");
+        } else if (task == "FRONTIER_ORBITALS") {
+            std::vector<int> indsa0;
+            std::vector<int> indsb0;
+            std::vector<std::string> labelsa;
+            std::vector<std::string> labelsb;
+            CharacterTable ct = basisset_->molecule()->point_group()->char_table();
+            if (nalpha_ == nbeta_) {
+                indsa0.push_back(nalpha_-1);
+                labelsa.push_back(to_string(std::get<1>(info_a_[nalpha_-1]) + 1) + "-" +
```
Comment (dgasmith) [inline psi4/src/psi4/libcubeprop/cubeprop.cc]: We can use `std::to_string` these days. Should probably try to kill off the old Psi `to_string` at some point.

File: psi4/src/psi4/libcubeprop/csg.cc [hunk 5]
```diff
@@ -649,6 +649,33 @@ void CubicScalarGrid::compute_orbitals(std::shared_ptr<Matrix> C, const std::vec
...
+    for (int k = 0; k < indices.size(); k++) {
+        C_DCOPY(primary_->nbf(), &Cp[0][indices[k]], C->colspi()[0], &C2p[0][k], C2->colspi()[0]);
+    }
+    double** v_t = block_matrix(indices.size(), npoints_);
+    double*  v = new double[npoints_];
+    memset(v_t[0], '\0', indices.size() * npoints_ * sizeof(double));
+    memset(v, '\0', npoints_ * sizeof(double));
+    add_orbitals(v_t, C2);
+    for (int i = 0; i < npoints_; i++) {
+        if (square) v[i] = (v_t[0][i] - v_t[1][i])*(v_t[0][i] + v_t[1][i]);
```
Comment (dgasmith) [inline psi4/src/psi4/libcubeprop/csg.cc]: Great place to use brackets `if (square) { ... `.

File: psi4/src/psi4/libcubeprop/csg.cc [hunk 6]
```diff
@@ -649,6 +649,33 @@ void CubicScalarGrid::compute_orbitals(std::shared_ptr<Matrix> C, const std::vec
     }
     free_block(v);
 }
+void CubicScalarGrid::compute_difference(std::shared_ptr<Matrix> C, const std::vector<int>& indices,
+                                         const std::string& label, bool square, const std::string& type) {
```
Comment (dgasmith) [inline psi4/src/psi4/libcubeprop/csg.cc]: Best to make this `const bool square`, the compiler will likely optimize the loop below.

File: psi4/src/psi4/libcubeprop/cubeprop.cc [hunk 7]
```diff
@@ -154,6 +158,52 @@ void CubeProperties::raw_compute_properties() {
...
+                indsa0.push_back(nalpha_-2);
+                labelsa.push_back(to_string(std::get<1>(info_a_[nalpha_-2]) + 1) + "-" +
+                                  ct.gamma(std::get<2>(info_a_[nalpha_-2])).symbol() + "_DOMO");
+                indsb0.push_back(nbeta_-1);
+                labelsb.push_back(to_string(std::get<1>(info_b_[nbeta_-1]) + 1) + "-" +
+                                  ct.gamma(std::get<2>(info_b_[nbeta_-1])).symbol() + "_DOMO");
+            }
+            if (indsa0.size()) compute_orbitals(Ca_, indsa0, labelsa, "Psi_a");
+            if (indsb0.size()) compute_orbitals(Cb_, indsb0, labelsb, "Psi_b");
+        } else if (task == "DUAL_DESCRIPTOR") {
```
Comment (loriab) [inline psi4/src/psi4/libcubeprop/cubeprop.cc]: Add the DOI links here perhaps.

**Comment by fevangelista** [comment]
Really love the "FRONTIER_ORBITALS" feature. Another one that might be nice to have is "ACTIVE_ORBITALS" for printing active orbitals in CASSCF computations.

**Comment by loriab** [comment]
Would be good to add new options [here](https://github.com/psi4/psi4/blob/master/psi4/src/read_options.cc#L209-L217) and [here](https://github.com/psi4/psi4/blob/master/doc/sphinxman/source/cubeprop.rst#cubeprop-tasks) which turns into [manual](http://psicode.org/psi4manual/master/cubeprop.html?highlight=basis_functions#cubeprop-tasks)

File: psi4/src/psi4/libcubeprop/csg.cc (hunk 3 above)
Comment (PeterKraus) [reply psi4/src/psi4/libcubeprop/csg.cc]: I took the `compute_orbitals` above as a guide. Is this symmetry check necessary, and if yes, shouldn't it be done for the orbitals as well?

File: psi4/src/psi4/libcubeprop/cubeprop.cc (hunk 7 above)
Comment (PeterKraus) [reply psi4/src/psi4/libcubeprop/cubeprop.cc]: I've added a short comment here and tried to describe the new additions in the docs.

**Comment by PeterKraus** [comment]
@fevangelista I am absolutely clueless about CAS methods, so I'd rather not go there.

@TermeHansen If I understand correctly, you'd like a `"VALENCE_DENSITY"` keyword, which plots the density due to just the valence electrons? I don't think it's terribly hard to implement, but I'd rather do that in a separate PR.

**Comment by TermeHansen** [comment]
That is fine. I have some test cases with chargemol where I believe it can be very useful. Please ping me on the PR

**Comment by loriab** [comment]
@PeterKraus, afraid you'll have to rebase to get #1198 in to get Travis to pass.

**Comment by dgasmith** [comment]
This looks good to me, but does need some tests. Examples can be found i `cubeprop/*`, I would borrow from these.

**Comment by dgasmith** [comment]
Sorry to be nitpick, but could you call the test `cubeprop-frontier` to describe the type of test rather than the molecule used? This will be good to go then.

**Pull request merged**